WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,202 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.Keys;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using WalletWasabi.WabiSabi.Crypto;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class CoinJoinClient : BackgroundService, IDisposable
+	{
+		private bool _disposedValue;
+		private CredentialPool AmountCredentialPool { get; } = new();
+		private CredentialPool VsizeCredentialPool { get; } = new();
+		private ClientRound Round { get; }
+		public IArenaRequestHandler ArenaRequestHandler { get; }
+		public Kitchen Kitchen { get; }
+		public KeyManager Keymanager { get; }
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private Coin[] Coins { get; set; }
+		private Random Random { get; } = new();
+
+		public CoinJoinClient(
+			ClientRound round,
+			IArenaRequestHandler arenaRequestHandler,
+			IEnumerable<Coin> coins,
+			Kitchen kitchen,
+			KeyManager keymanager)
+		{
+			Round = round;
+			ArenaRequestHandler = arenaRequestHandler;
+			Kitchen = kitchen;
+			Keymanager = keymanager;
+			SecureRandom = new SecureRandom();
+			Coins = coins.ToArray();
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			try
+			{
+				// Register coins.
+				AliceClient[] aliceClients = await RegisterCoinsAsync(stoppingToken).ConfigureAwait(false);
+
+				// Confirm coins.
+				await ConfirmConnectionsAsync(aliceClients, stoppingToken).ConfigureAwait(false);
+
+				// Planning
+				var outputs = DecomposeAmounts(stoppingToken);
+
+				// Output registration.
+				await ReissueAndRegisterOutputsAsync(outputs, stoppingToken).ConfigureAwait(false);
```

Comment:
For this flow with regards to #5726, if you want to integrate the planner, and assuming you're happy with the API, then I think you can do the following:

`DecomposeAmounts` should happen first thing. The input values, the output values as well as the corresponding vsizes for each should then be combined to an arbitrarily ordered list of amounts, specifying negative values for outputs.

This list of values given to `DependencyGraph.ResolveCredentialDependencies(IEnumerable<IEnumerable<long>>)`, the inner enumerable is of length 2 for both credential types, the first `long` is the amount value and the second is the vsize (numbers should correspond to the balance proofs, i.e. effective values).

The returned value is a resolved graph, with the nodes for the given values appearing in order (and node IDs are numerical ascending), and the additional nodes in the `Vertices` property describe any reissuance requests that need to be executed. Indexing vertices in the same order as the input argument yields the corresponding nodes.

The related nodes for reissuance requests can be discovered by either iterating the remaining vertices in the property, or by enumerating the edges with `InEdges` or `OutEdges` method, every edge is a `CredentialDependency` with a type enum and a value.

In order to provide the Alice clients with the credential amounts to use at connection confirmation, enumerate the `OutEdges` of the nodes with a positive `InitialBalance`. The corresponding `CredentialDependency` refer to the target request, which may either be a reissuance request (`InitialBalance(credentialType) == 0`) or an output registration (`InitialBalance(credentialType) < 0).

Doing this traversal is probably easier than than iterating over the vertices because traversing the graph breadth first presents them in a dependency satisfying order, making it easier to generate a corresponding DAG of tasks. For each node there would to be a task, basically doing the same thing as the channel read in @lontivero's #5716, the part that waits on all of the tasks it depends on, then gives the client its credentials, and when the client finishes the response credentials are then pushed into the tasks of the dependent clients. It's not clear to me if this should be done by the credential pool or the round client.

The number of zero credentials needed for each request node is given by `graph.InDegree` or `graph.OutDegree`, but not directly as this value doesn't account for the various special cases.

Reply:
I cannot add/comment on this right now. I need to see a test, how this is implemented IRL.